janus_core/helpers/correlator.py

```diff
@@ -0,0 +1,287 @@
+    def _add(self, a: float, b: float, block: int):
+        """
+        Propagate update down block hierarchy.
+
+        Parameters
+        ----------
+        a : float
+            Newly observed value of left correland/average.
+        b : float
+            Newly observed value of right correland/average.
+        block : int
+            Block in the hierachy being updated.
+        """
+        if block == self._blocks:
+            return
+
+        shift = self._shift_index[block]
+        self._max_block_used = max(self._max_block_used, block)
+        self._shift[block, shift, :] = a, b
+        self._accumulator[block, :] += a, b
+        self._shift_not_null[block, shift] = True
+        self._count_accumulated[block] += 1
+
+        if self._count_accumulated[block] == self._averaging:
+            self._add(
+                self._accumulator[block, 0] / self._averaging,
+                self._accumulator[block, 1] / self._averaging,
+                block + 1,
+            )
+            self._accumulator[block, :] = 0.0
+            self._count_accumulated[block] = 0
+
+        i = self._shift_index[block]
+        if block == 0:
+            j = i
+            for point in range(self._points):
+                if self._shifts_valid(block, i, j):
+                    self._correlation[block, point] += (
+                        self._shift[block, i, 0] * self._shift[block, j, 1]
+                    )
+                    self._count_correlated[block, point] += 1
+                j -= 1
+                if j < 0:
+                    j += self._points
+        else:
+            for point in range(self._min_dist, self._points):
+                if j < 0:
+                    j = j + self._points
+                if self._shifts_valid(block, i, j):
+                    self._correlation[block, point] += (
+                        self._shift[block, i, 0] * self._shift[block, j, 1]
+                    )
+                    self._count_correlated[block, point] += 1
+                j = j - 1
+        self._shift_index[block] = (self._shift_index[block] + 1) % self._points
+
+    def _shifts_valid(self, block: int, p_i: int, p_j: int) -> bool:
+        """
+        True if the shift registers have data.
+
+        Parameters
+        ----------
+        block : int
+            Block to check the shift register of.
+        p_i : int
+            Index i in the shift (left correland).
+        p_j : int
+            Index j in the shift (right correland).
+
+        Returns
+        -------
+        bool
+            Whether the shift indices have data.
+        """
+        return self._shift_not_null[block, p_i] and self._shift_not_null[block, p_j]
+
+    def get(self) -> tuple[Iterable[float], Iterable[float]]:
+        """
+        Obtain the correlation and lag times.
+
+        Returns
+        -------
+        tuple[Iterable[float], Iterable[float]]
+            The correlation and lags.
```

Comment:
Under numpy docs you can address the tuple elements independently by a name like you would for parameters.